### PR TITLE
Use label translator to translate groups and tabs

### PR DESCRIPTION
**Pull request by EmmanuelVella**
New PR from https://github.com/sonata-project/SonataAdminBundle/pull/2870 for 3.x branch
## Changelog

``` markdown
### Fixed
- Fixed not translated group and tabs
```
## Subject

Currently, when calling ->with() or ->tab() on a form or show mapper, the label is the exact name you passed : ->with('myTest') would give the myTest label. This force us to eventually pass a manual label like ->with('form.group_my_test') to get the good output.

With this PR, the label is created just like in the add method. So ->with('myTest') would give the form.group_my_test label.

### CODE REVIEW

File: Form/FormMapper.php [hunk 1]
```diff
@@ -301,4 +301,12 @@ protected function setTabs(array $tabs)
     {
         $this->admin->setFormTabs($tabs);
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getName()
+    {
+        return 'form';
```
Comment (greg0ire) [inline Form/FormMapper.php]: This is not forward-compatible, you have to return a class name for the last versions of Symfony.

**Comment by EmmanuelVella** [comment]
**Currently, there is no BC for group/tab name containing no dot**
```
$form->with('myGroupName')
```
will output `form.group_my_group_name`

If we need BC at this level, it would need : 
- add a config parameter until next major release
- ...

Please feel free to suggest your ideas

File: Form/FormMapper.php (hunk 1 above)
Comment (EmmanuelVella) [inline Form/FormMapper.php]: Ok, fixing it !

File: Mapper/BaseGroupedMapper.php [hunk 2]
```diff
@@ -256,6 +262,16 @@ public function hasOpenTab()
     abstract protected function setTabs(array $tabs);
 
     /**
+     * TODO: make this method abstract before releasing 4.0
```
Comment (greg0ire) [inline Mapper/BaseGroupedMapper.php]: Replace `TODO` with `NEXT_MAJOR` please

File: Form/FormMapper.php (hunk 1 above)
Comment (EmmanuelVella) [inline Form/FormMapper.php]: Hum are you sure about this ? This is only called here : https://github.com/sonata-project/SonataAdminBundle/pull/4028/files#diff-d561f27bd57cd565613d211d8d8bb0afR73 and this is not related to symfony ``AbstractType``

Comment (greg0ire) [inline Form/FormMapper.php]: Oh right ok, my bad, the `form` part confused me.

File: Mapper/BaseGroupedMapper.php [hunk 3]
```diff
@@ -70,11 +70,17 @@ public function with($name, array $options = array())
...
+            'label' => $this->admin->getLabelTranslatorStrategy()->getLabel($name, $this->getName(), 'group'),
             'translation_domain' => null,
             'name' => $name,
             'box_class' => 'box box-primary',
         );
 
+        // BC layer for manual label
+        if (strpos($name, '.')) {
+            $defaultOptions['label'] = $name;
+        }
```
Comment (greg0ire) [inline Mapper/BaseGroupedMapper.php]: I think you should go on with your idea of a parameter, and remove this, the config parameter is the only clean solution I can imagine.

So: 
- in 3.x `translate_group_label`: defaults to false, and raises a deprecation warning in that case.
- In 4.x : completely ignored, raises a deprecation warning when set
- In 5.x: removed

File: Resources/views/CRUD/base_edit_form.html.twig [hunk 4]
```diff
@@ -35,7 +35,12 @@
                         <div class="nav-tabs-custom">
                             <ul class="nav nav-tabs" role="tablist">
                                 {% for name, form_tab in admin.formtabs %}
-                                    <li{% if loop.index == 1 %} class="active"{% endif %}><a href="#tab_{{ admin.uniqid }}_{{ loop.index }}" data-toggle="tab"><i class="fa fa-exclamation-circle has-errors hide"></i> {{ admin.trans(name, {}, form_tab.translation_domain) }}</a></li>
+                                    <li{% if loop.first %} class="active"{% endif %}>
+                                        <a href="#tab_{{ admin.uniqid }}_{{ loop.index }}" data-toggle="tab">
+                                            <i class="fa fa-exclamation-circle has-errors hide"></i>
+                                            {{ admin.trans(form_tab.label, {}, form_tab.translation_domain) }}
```
Comment (greg0ire) [inline Resources/views/CRUD/base_edit_form.html.twig]: For those wondering, the thing that changed here is `form_tab.label` (is there something else @EmmanuelVella ?). Please next time do separate commits with `git add --patch`, otherwise it's quite hard to read.

File: Mapper/BaseGroupedMapper.php [hunk 5]
```diff
@@ -256,6 +262,16 @@ public function hasOpenTab()
     abstract protected function setTabs(array $tabs);
 
     /**
+     * TODO: make this method abstract before releasing 4.0
+     *
+     * @return string
+     */
+    /*abstract*/ protected function getName()
+    {
+        return 'default';
```
Comment (greg0ire) [inline Mapper/BaseGroupedMapper.php]: You have to raise a deprecation warning here saying that getName() should be implemented if you're going to make this abstract.

Comment (greg0ire) [inline Mapper/BaseGroupedMapper.php]: Also, can you clarify what this has to do with the issue at hand?

**Comment by EmmanuelVella** [comment]
@greg0ire I will work again on this PR soon, too busy ATM !

**Comment by greg0ire** [comment]
Rebasing should fix the build, see #4079

**Comment by OskarStark** [comment]
@greg0ire I let you merge this PR after a final review.

From my POV everything is fine

**Comment by OskarStark** [comment]
`minor` or `patch` ?

File: Mapper/BaseGroupedMapper.php (hunk 3 above)
Comment (greg0ire) [inline Mapper/BaseGroupedMapper.php]: Hello?

File: Resources/views/CRUD/base_edit_form.html.twig (hunk 4 above)
Comment (EmmanuelVella) [inline Resources/views/CRUD/base_edit_form.html.twig]: Fixed

File: Mapper/BaseGroupedMapper.php (hunk 5 above)
Comment (EmmanuelVella) [inline Mapper/BaseGroupedMapper.php]: I just added a deprecation warning.

This method is used to build the translation id : ``form.group_foo``, ``show.group_foo`` or ``default.group_foo``.

If I return nothing, it will only give ``group_foo``.

File: Mapper/BaseGroupedMapper.php (hunk 3 above)
Comment (EmmanuelVella) [inline Mapper/BaseGroupedMapper.php]: Ok I just added the config parameter and use it here. However, has the `BaseGroupedMapper` class is not a service, I rely on this call : https://github.com/sonata-project/SonataAdminBundle/pull/4028/files#diff-d561f27bd57cd565613d211d8d8bb0afR80

I also can inject this parameter when creating the `FormMapper`/`ShowMapper` : 
- via a constructor argument : will make the class not BC
- via a setter : may not be set

What do you think ?

File: Mapper/BaseGroupedMapper.php [hunk 6]
```diff
@@ -70,11 +70,17 @@ public function with($name, array $options = array())
             'collapsed' => false,
             'class' => false,
             'description' => false,
+            'label' => $this->admin->getLabelTranslatorStrategy()->getLabel($name, $this->getName(), 'group'),
             'translation_domain' => null,
```
Comment (core23) [inline Mapper/BaseGroupedMapper.php]: What about setting the `translation_domain` to `false`? This would follow symfony standard.

File: Mapper/BaseGroupedMapper.php (hunk 3 above)
Comment (greg0ire) [inline Mapper/BaseGroupedMapper.php]: I think there is another black magic solution : use `func_get_args()`. Never occurred to me until now but that might be a good way to do BC hacks without modifying method signatures.

Comment (EmmanuelVella) [inline Mapper/BaseGroupedMapper.php]: Could you give me an example please ?

Comment (EmmanuelVella) [inline Mapper/BaseGroupedMapper.php]: Ho ok I got it !

Comment (greg0ire) [inline Mapper/BaseGroupedMapper.php]: [Sure](https://3v4l.org/MZpWl)

Comment (EmmanuelVella) [inline Mapper/BaseGroupedMapper.php]: This is the same thing than using setter injection, no ?

Comment (greg0ire) [inline Mapper/BaseGroupedMapper.php]: > Ho ok I got it !

Great! It's a bit ugly but you get injection and no BC-break, and it's a piece of code that will be removed so I think it's a good trade-off

Comment (greg0ire) [inline Mapper/BaseGroupedMapper.php]: WDYT @sonata-project/contributors ?

Comment (greg0ire) [inline Mapper/BaseGroupedMapper.php]: > This is the same thing than using setter injection, no ?

Not really since we won't have to remove a setter. Plus you can check whether the argument is provided or not in the constructor, and throw an exception if it isn't.

Comment (EmmanuelVella) [inline Mapper/BaseGroupedMapper.php]: Ok, however it will only deport the problem in the `AbstractAdmin` class, which initialize the `FormMapper` and the `ShowMapper` : https://github.com/sonata-project/SonataAdminBundle/blob/3.x/Admin/AbstractAdmin.php#L1265

Comment (greg0ire) [inline Mapper/BaseGroupedMapper.php]: Maybe it's not worth the hassle then. Plus I say I'm not changing the API, but in fact, I am, just in a hackish, hidden way. Maybe you should just leave it as is. Keep it mind that this dirty hack will disappear on next major, so, not that big of a deal.

**Comment by core23** [comment]
Please rebase, so we can merge this @EmmanuelVella

**Comment by core23** [comment]
> `minor` or `patch` ?

Minor, becaue we introduce a new feature.

File: Mapper/BaseGroupedMapper.php [hunk 7]
```diff
@@ -70,11 +70,17 @@ public function with($name, array $options = array())
...
             'name' => $name,
             'box_class' => 'box box-primary',
         );
 
+        // BC layer for manual label
```
Comment (core23) [inline Mapper/BaseGroupedMapper.php]: Is this a BC layer? Will we remove this behavior on the next major?

**Comment by OskarStark** [comment]
@EmmanuelVella can you please try to finish this PR? 🌷

**Comment by EmmanuelVella** [comment]
@OskarStark Hey, I rebased and fixed the tests. Sorry for the delay !

**Comment by OskarStark** [comment]
not a big deal, thank you :)

**Comment by greg0ire** [comment]
Thanks @EmmanuelVella !

**Comment by OskarStark** [comment]
This one is released: https://github.com/sonata-project/SonataAdminBundle/releases/tag/3.20.0

**Comment by fracsi** [comment]
In the `base_show.html.twig` template a there is an undefined variable `view_group`: https://github.com/sonata-project/SonataAdminBundle/blob/3.20.0/Resources/views/CRUD/base_show.html.twig#L47

It should be: `show_tab.label`

**Comment by OskarStark** [comment]
@fracsi can you please open a bug ticket?

**Comment by OskarStark** [comment]
@EmmanuelVella can you please take care of #4028?
